### ISSUE TITLE
Add quota pre-check to analyses that consume quota

### ISSUE THREAD

**Issue by rafatower** [CartoDB/camshaft#238]
Now we have a couple of functions to check quota before executing bulk operations on the DS API:

```sql
select * from cdb_dataservices_client.cdb_service_quota_info();
    service     | monthly_quota | used_quota | soft_limit |     provider     
----------------+---------------+------------+------------+------------------
 isolines       |          1000 |          0 | f          | heremaps
 hires_geocoder |             5 |          2 | f          | mapzen
 routing        |       1500000 |          0 | f          | mapzen
 observatory    |          1000 |          0 | f          | data observatory
(4 rows)
```

and

```sql
select cdb_dataservices_client.cdb_enough_quota('isolines', 500);
 cdb_enough_quota 
------------------
 t
(1 row)

select cdb_dataservices_client.cdb_enough_quota('isolines', 5000);
 cdb_enough_quota 
------------------
 f
(1 row)
```

This way we shall avoid launching analyses that we know upfront that will raise a quota exhausted exception.

To some extend related to the more general problem described here: https://github.com/CartoDB/dataservices-api/issues/314

cc/ @xavijam , this is the place to go for quota checks.

Update 1: This is the PR for the FE side: https://github.com/CartoDB/cartodb/pull/11480
Update 2: ~~Blocked by this other issue~~: https://github.com/CartoDB/camshaft/issues/260
Update 3: Unblocked \o/

**Comment by rafatower** [comment]
Related: https://github.com/CartoDB/cartodb/issues/9231

**Comment by saleiva** [comment]
Adding to BUILDER kanban to refresh the UI

**Comment by nobuti** [comment]
In the userModel right now we have some data exposed related to analyses. For example:
```
here_isolines: {
    block_price: null
    hard_limit: true
    monthly_use: 0
    quota: 0
}
```
And what we get calling the data service API is something like this:
```
{
    monthly_quota: 0,
    provider: "heremaps",
    service: "isolines",
    soft_limit: false,
    used_quota: 0
}
```
It seems there is some duplicity in the data. So my questions are:
- the translation between both objects are:
    - monthly_quota === quota
    - soft_limit === !hard_limit
    - used_quota === monthly_use
- should we stop exposing those attributes (quota, hard_limit, monthly_use) as part of the user data payload?

**Comment by rafatower** [comment]
> should we stop exposing those attributes (quota, hard_limit, monthly_use) as part of the user data payload?

A certain amount of duplication can be hardly avoided because of how syncing work: central -> boxes -> DS api.

But when it comes to the user data payload, I'd say it's best to avoid that duplication, if possible. That part of the model comes from pre-api times.

**Comment by xavijam** [comment]
Hey @rafatower, why do we have to pass the number of rows in the function?. Couldn't we just pass the query in order to have the api calculate that?.
IMO, that will prevent us to make two requests (one calculating the number of rows) and another to get the "quota" details.
Not sure if I should ping any other dev to this question.

**Comment by rafatower** [comment]
@xavijam there are two major reasons:
- The Dataservices API shall not be coupled with the User DB
- Certain analyses (at present, isolines) can consume more than 1 credit/row (when you request more than one tract). https://carto.com/docs/carto-engine/dataservices-api/quota-information#cdbenoughquotaservice-text-inputsize-numeric

Not 100% sure about the best place to put such a function to perform just 1 query, but I'm also concerned about coupling a simple quota info api with the specifics of the services. Shouldn't we consider that an improvement that can be deferred?

**Comment by xavijam** [comment]
Hey, as @javisantana suggested last day, we will try to get a rows estimation (https://www.postgresql.org/docs/8.3/static/row-estimation-examples.html) and then check the quota.
But, as @saleiva agreed, it is likely we need to change the API in order to let the user run the analysis although he/she doesn't have enough credits, do we?.

**Comment by rafatower** [comment]
There's `soft_limit` and it's configurable through Central. Shall we stick to it rather than changing the API?

**Comment by xavijam** [comment]
Well, isn't it the same right?. You could want to run the analysis until you finish your credits (no matter if it completes the whole "dataset") and not be charged. The `soft_limit` is great for people that don't want surprises runnning analyses with quotas.

**Comment by rafatower** [comment]
To sum up from our mini-meeting:

- In case there's no enough quota: the analysis layer does up to the quota allowance, then returns NULL's when exhausted.
- If the `soft_limit` is set to true, it will continue passed the quota allowance, returning valid values from DS and then billing the overquota as such.

E.g: think of a geocoding analysis

| monthly_quota | used_quota | no. of rows to geocode | soft_limit | result |
| -------------------- | --------------- | ------------ | -------- | -------- |
| 100 | 50 | 200 | false | 50 rows geocoded and 150 with NULL's. |
| 100 | 50 | 200 | true  | 200 rows geocoded, overquota of 150 recorded. |

- These changes provides visibility on what a user can expect before triggering an analysis
- An ideal API would be more similar to what we had for the Editor geocoder: ask for an estimation of price. That is out of the scope of this ticket. The right place to implement that is not the DS API as it shall not know about price plans and the likes. We'll defer this to a later project.

HTH

**Comment by nobuti** [comment]
@rafatower one question, according to the documentation:

> Note that some services consume more than one credit per row/call. For example, isolines (with more than one range/track) consume (N rows x M ranges) credits; indicating that the input size should be N x M.

Does it mean that we need to provide the right input including ranges or may we assume the API takes care of the ranges? Does this happen with another kind of analysis?

**Comment by rafatower** [comment]
You need to provide the right input, DS API cannot know how many tracks you are going to request next. Calls are independent from one another.

**Comment by arianaescobar** [comment]
Hey guys, @saleiva asked me to share this with you: 

![quota warnings](https://cloud.githubusercontent.com/assets/16247628/22514946/3f5888f4-e8a1-11e6-83a0-1f110c7b1ea8.png)

If you have any question about it, let me know :)

**Comment by nobuti** [comment]
@rafatower one question, does `cdb_enough_quota` take into account the `soft_limit`? I can see nothing in the documentation but I'm getting `true` no matter the quota is if `soft_limit` is `true`:

![screen shot 2017-02-10 at 13 17 41](https://cloud.githubusercontent.com/assets/1366843/22826578/75d85078-ef93-11e6-8eb0-2cc50f821313.png)

![screen shot 2017-02-10 at 13 18 12](https://cloud.githubusercontent.com/assets/1366843/22826585/7a68f49e-ef93-11e6-847f-f1bc3d53087b.png)

Is that ok?

**Comment by rafatower** [comment]
`soft_limit == true` means that you can geocode or use whatever service past your quota allowance, so that's correct.

**Comment by javierarce** [comment]
@nobuti this issue is in CR but still blocked… shall I wait to do the review?

**Comment by nobuti** [comment]
@javierarce yes please, I'm afraid some work will be done once we find a workaround using `EXPLAIN` in sql in order to get an estimation.

**Comment by xavijam** [comment]
Hey guys, sorry to ask, but how is it going?

**Comment by rafatower** [comment]
@xavijam I'm deploying the [changes in camshaft](https://github.com/CartoDB/camshaft/issues/260) right now, that should unblock this ticket.

**Comment by xavijam** [comment]
The master!

**Comment by xavijam** [comment]
Status: 
- Seems like Builder implementation is ready. We have to make a final pass if necessary.
- We found a problem with quotas described here -> https://github.com/CartoDB/dataservices-api/issues/349 (pending final decision if it is blocker or not @noguerol @saleiva ).

cc @javisantana @rochoa @rafatower

**Comment by noguerol** [comment]
Not a blocker, we move forward.

**Comment by xavijam** [comment]
FINALLY DEPLOYED! 👏  great job all!

**Comment by xavijam** [comment]
Hey, no, let's not reopen this one, let's create a new one. Enough thread for this feature... :(
